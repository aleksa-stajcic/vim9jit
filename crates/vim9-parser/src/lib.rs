#![allow(dead_code)]

use anyhow::Result;
use vim9_lexer::new_lexer;
use vim9_lexer::Lexer;
use vim9_lexer::Token;
use vim9_lexer::TokenKind;

#[derive(Debug)]
pub struct Program {
    commands: Vec<ExCommand>,
}

#[derive(Debug, PartialEq)]
pub enum ExCommand {
    Vim9Script(Vim9ScriptCommand),
    Var(VarCommand),
    Return(ReturnCommand),
    Def(DefCommand),
    If(IfCommand),
    Call(CallCommand),
    Finish(Token),

    Skip,
    EndOfFile,
    Comment(Token),
    NoOp(Token),
}

#[derive(Debug, PartialEq)]
pub struct Vim9ScriptCommand {
    noclear: bool,
    eol: Token,
}

#[derive(Debug, PartialEq)]
pub struct VarCommand {
    var: Token,
    pub name: Name,
    equal: Token,
    pub expr: Expression,
    eol: Token,
}

impl VarCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        let var = VarCommand {
            var: parser.expect_token_with_text(TokenKind::Identifier, "var")?,
            name: Name::parse(parser)?,
            // TODO: Type Hints
            equal: parser.expect_token(TokenKind::Equal)?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        };

        Ok(ExCommand::Var(var))
    }
}

#[derive(Debug, PartialEq)]
pub struct Body {
    pub commands: Vec<ExCommand>,
}

impl Body {
    pub fn parse_until(parser: &mut Parser, identifier: &str) -> Result<Body> {
        let mut commands = vec![];
        while parser.current_token.text != identifier {
            commands.push(parser.parse_command()?);
        }

        Ok(Body { commands })
    }
}

#[derive(Debug, PartialEq)]
pub struct IfCommand {
    if_tok: Token,
    pub condition: Expression,
    if_eol: Token,
    pub body: Body,
    endif_tok: Token,
    endif_eol: Token,
}

impl IfCommand {
    pub fn parser(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::If(IfCommand {
            if_tok: parser.expect_identifier_with_text("if")?,
            condition: Expression::parse(parser)?,
            if_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "endif")?,
            endif_tok: parser.expect_identifier_with_text("endif")?,
            endif_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct DefCommand {
    def: Token,
    pub name: Name,
    pub args: Signature,
    pub ret: Option<Type>,
    def_eol: Token,
    pub body: Body,
    enddef: Token,
    end_eol: Token,
}

impl DefCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Def(DefCommand {
            def: parser.expect_identifier_with_text("def")?,
            name: Name::parse(parser)?,
            args: Signature::parse(parser)?,
            ret: {
                if parser.current_token.kind == TokenKind::Colon {
                    Some(Type {
                        colon: parser.expect_token(TokenKind::Colon)?,
                        name: Name::parse(parser)?,
                    })
                } else {
                    None
                }
            },
            def_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "enddef")?,
            enddef: parser.expect_identifier_with_text("enddef")?,
            end_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct CallCommand {
    call: Option<Token>,
    pub name: Name,
    pub args: Arguments,
    eol: Token,
}

impl CallCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Call(CallCommand {
            call: None,
            name: Name::parse(parser)?,
            args: Arguments::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct Arguments {
    open: Token,
    pub args: Vec<Expression>,
    close: Token,
}

impl Arguments {
    pub fn parse(parser: &mut Parser) -> Result<Arguments> {
        Ok(Arguments {
            open: parser.expect_token(TokenKind::LeftParen)?,
            args: {
                let mut args = vec![];
                while parser.current_token.kind != TokenKind::RightParen {
                    args.push(Expression::parse(parser)?);

                    // TODO: Consume ','?
                }

                args
            },
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Type {
    colon: Token,
    pub name: Name,
}

#[derive(Debug, PartialEq)]
pub struct Signature {
    open: Token,
    close: Token,
}

impl Signature {
    fn parse(parser: &mut Parser) -> Result<Signature> {
        Ok(Self {
            open: parser.expect_token(TokenKind::LeftParen)?,
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Parameter {
    pub name: Name,
    colon: Token,
    // pub typ: ,
}

#[derive(Debug, PartialEq)]
pub struct ReturnCommand {
    ret: Token,
    pub expr: Expression,
    eol: Token,
}

impl ReturnCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Return(Self {
            ret: parser.expect_token_with_text(TokenKind::Identifier, "return")?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Name {
    Identifier(RawIdentifier),
    ScopedIdentifier,
}

impl Name {
    fn parse(parser: &mut Parser) -> Result<Name> {
        // Todo: Other names
        Ok(Name::Identifier(RawIdentifier {
            name: parser.expect_token(TokenKind::Identifier)?.text,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Expression {
    Name,
    Number(Number),
    String(VimString),
}

impl Expression {
    pub fn parse(parser: &mut Parser) -> Result<Expression> {
        let expr = match &parser.current_token.kind {
            TokenKind::Integer => Expression::Number(Number {
                value: parser.current_token.text.clone(),
            }),
            TokenKind::DoubleQuoteString => {
                Expression::String(VimString::DoubleQuote(parser.current_token.text.clone()))
            }
            TokenKind::SingleQuoteString => {
                Expression::String(VimString::SingleQuote(parser.current_token.text.clone()))
            }

            _ => todo!("expresson parse: {:?}", parser.current_token),
        };

        parser.next_token();
        Ok(expr)
    }
}

#[derive(Debug, PartialEq)]
pub enum VimString {
    SingleQuote(String),
    DoubleQuote(String),
}

#[derive(Debug, PartialEq)]
pub struct RawIdentifier {
    name: String,
}

#[derive(Debug, PartialEq)]
pub struct Number {
    // ??? should this be a number?
    value: String,
}

#[derive(Debug)]
pub struct Parser {
    lexer: Lexer,
    current_token: Token,
    peek_token: Token,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        Self {
            current_token: lexer.next_token(),
            peek_token: lexer.next_token(),
            lexer,
        }
    }

    pub fn expect_eol(&mut self) -> Result<Token> {
        self.expect_token(TokenKind::EndOfLine)
    }

    pub fn expect_token(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            return Err(anyhow::anyhow!("Got token: {:?}, Expected: {:?}", token, kind));
        }

        self.next_token();
        Ok(token)
    }

    pub fn expect_token_with_text(&mut self, kind: TokenKind, text: &str) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            panic!("Got token: {:?}, Expected kind: {:?}", token, kind);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected kind: {:?}", token, kind));
        }

        if token.text != text {
            panic!("Got token: {:?}, Expected text: {:?}", token, text);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected text: {:?}", token, text));
        }

        self.next_token();
        Ok(token)
    }

    pub fn peek_identifier_with_text(&self, text: &str) -> bool {
        self.peek_token.kind == TokenKind::Identifier && self.peek_token.text == text
    }

    pub fn expect_identifier_with_text(&mut self, text: &str) -> Result<Token> {
        self.expect_token_with_text(TokenKind::Identifier, text)
    }

    pub fn next_token(&mut self) -> Token {
        self.current_token = self.peek_token.clone();
        self.peek_token = self.lexer.next_token();

        self.current_token.clone()
    }

    fn command_match(&self, full: &str) -> bool {
        self.current_token.text == full
    }

    pub fn parse_command(&mut self) -> Result<ExCommand> {
        // For the following branches, you need to return early if it completely consumes
        // the last character and advances past.
        //
        // This is the desired behavior for `parse` which will consume until the end of line
        // generally speaking.
        let ex = match &self.current_token.kind {
            TokenKind::EndOfFile => panic!("EndOfFile!"),
            TokenKind::Comment => ExCommand::Comment(self.current_token.clone()),
            TokenKind::EndOfLine => ExCommand::NoOp(self.current_token.clone()),
            TokenKind::Identifier => {
                if self.command_match("vim9script") {
                    self.next_token();
                    ExCommand::Vim9Script(Vim9ScriptCommand {
                        noclear: if self.current_token.kind == TokenKind::EndOfLine {
                            false
                        } else {
                            self.expect_identifier_with_text("noclear")?;
                            true
                        },
                        eol: self.expect_eol()?,
                    })
                } else if self.command_match("var") {
                    return Ok(VarCommand::parse(self)?);
                } else if self.command_match("def") {
                    return Ok(DefCommand::parse(self)?);
                } else if self.command_match("return") {
                    return Ok(ReturnCommand::parse(self)?);
                } else if self.command_match("if") {
                    return Ok(IfCommand::parser(self)?);
                } else if self.command_match("finish") {
                    ExCommand::Finish(self.current_token.clone())
                } else {
                    if self.peek_token.kind == TokenKind::LeftParen {
                        return Ok(CallCommand::parse(self)?);
                    } else {
                        ExCommand::NoOp(self.current_token.clone())
                    }
                }
            }
            _ => ExCommand::NoOp(self.current_token.clone()),
        };

        self.next_token();
        Ok(ex)
    }

    pub fn parse_program(&mut self) -> Program {
        let mut program = Program { commands: vec![] };

        while self.current_token.kind != TokenKind::EndOfFile {
            let command = self.parse_command().unwrap();
            if command != ExCommand::Skip {
                program.commands.push(command);
            }
        }

        program
    }
}

fn snapshot_parsing(input: &str) -> String {
    let lexer = new_lexer(input);
    let mut parser = Parser::new(lexer);
    let program = parser.parse_program();

    format!("{:#?}", program.commands)
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents));
                });
            }
        };
    }

    snapshot!(test_var, "../testdata/snapshots/simple_var.vim");
    snapshot!(test_ret, "../testdata/snapshots/simple_ret.vim");
    snapshot!(test_comment, "../testdata/snapshots/comment.vim");
    snapshot!(test_header, "../testdata/snapshots/header.vim");

    // TODO: Slowly but surely, we can work towards this
    // snapshot!(test_matchparen, "../testdata/snapshots/matchparen.vim");
}
