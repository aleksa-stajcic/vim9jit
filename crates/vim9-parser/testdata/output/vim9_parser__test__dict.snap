---
source: crates/vim9-parser/src/lib.rs
assertion_line: 1945
expression: snapshot_parsing(contents)
---
[
    Vim9Script(
        Vim9ScriptCommand {
            noclear: false,
            eol: Token(EndOfLine, "\n", (0,10)->(0,10)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (2,0)->(2,3)),
            ty: None,
            name: Raw(x),
            equal: Token(Equal, "=", (2,6)->(2,7)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (2,8)->(2,9)),
                    elements: [
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "hello", (2,10)->(2,15)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (2,15)->(2,17)),
                            value: String(
                                DoubleQuote(
                                    "world",
                                ),
                            ),
                            comma: Some(
                                Token(Comma, ",", (2,24)->(2,25)),
                            ),
                        },
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "another", (2,26)->(2,33)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (2,33)->(2,35)),
                            value: Boolean(
                                VimBoolean {
                                    token: Token(False, "false", (2,35)->(2,40)),
                                    value: false,
                                },
                            ),
                            comma: None,
                        },
                    ],
                    close: Token(RightBrace, "}", (2,41)->(2,42)),
                },
            ),
            eol: Token(EndOfLine, "\n", (2,42)->(2,42)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (3,0)->(3,3)),
            ty: None,
            name: Raw(y),
            equal: Token(Equal, "=", (3,6)->(3,7)),
            expr: Call(
                f: Identifier(Raw(len)) arg: [
                    Call(
                        f: Identifier(Raw(nvim_get_autcomds)) arg: [
                            Dict(
                                DictLiteral {
                                    open: Token(LeftBrace, "{", (3,30)->(3,31)),
                                    elements: [
                                        KeyValue {
                                            key: Literal(
                                                Literal {
                                                    token: Token(Identifier, "group", (3,31)->(3,36)),
                                                },
                                            ),
                                            colon: Token(SpacedColon, ": ", (3,36)->(3,38)),
                                            value: String(
                                                DoubleQuote(
                                                    "matchparen",
                                                ),
                                            ),
                                            comma: None,
                                        },
                                    ],
                                    close: Token(RightBrace, "}", (3,50)->(3,51)),
                                },
                            ),
                        ],
                    ),
                ],
            ),
            eol: Token(EndOfLine, "\n", (3,53)->(3,53)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (4,0)->(4,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (5,0)->(5,3)),
            ty: None,
            name: Raw(multiline),
            equal: Token(Equal, "=", (5,14)->(5,15)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (5,16)->(5,17)),
                    elements: [
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "x", (6,2)->(6,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (6,3)->(6,5)),
                            value: Identifier(
                                Raw(x),
                            ),
                            comma: None,
                        },
                    ],
                    close: Token(RightBrace, "}", (7,0)->(7,1)),
                },
            ),
            eol: Token(EndOfLine, "\n", (7,1)->(7,1)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (8,0)->(8,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (9,0)->(9,3)),
            ty: None,
            name: Raw(multiline_trail),
            equal: Token(Equal, "=", (9,20)->(9,21)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (9,22)->(9,23)),
                    elements: [
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "x", (10,2)->(10,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (10,3)->(10,5)),
                            value: Identifier(
                                Raw(x),
                            ),
                            comma: Some(
                                Token(Comma, ",", (10,6)->(10,7)),
                            ),
                        },
                    ],
                    close: Token(RightBrace, "}", (11,0)->(11,1)),
                },
            ),
            eol: Token(EndOfLine, "\n", (11,1)->(11,1)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (12,0)->(12,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (13,0)->(13,3)),
            ty: None,
            name: Raw(multiline_two),
            equal: Token(Equal, "=", (13,18)->(13,19)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (13,20)->(13,21)),
                    elements: [
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "x", (14,2)->(14,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (14,3)->(14,5)),
                            value: Identifier(
                                Raw(x),
                            ),
                            comma: Some(
                                Token(Comma, ",", (14,6)->(14,7)),
                            ),
                        },
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "y", (15,2)->(15,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (15,3)->(15,5)),
                            value: Identifier(
                                Raw(y),
                            ),
                            comma: None,
                        },
                    ],
                    close: Token(RightBrace, "}", (16,0)->(16,1)),
                },
            ),
            eol: Token(EndOfLine, "\n", (16,1)->(16,1)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (17,0)->(17,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (18,0)->(18,3)),
            ty: None,
            name: Raw(multiline_trail),
            equal: Token(Equal, "=", (18,20)->(18,21)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (18,22)->(18,23)),
                    elements: [
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "x", (19,2)->(19,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (19,3)->(19,5)),
                            value: Identifier(
                                Raw(x),
                            ),
                            comma: Some(
                                Token(Comma, ",", (19,6)->(19,7)),
                            ),
                        },
                        KeyValue {
                            key: Literal(
                                Literal {
                                    token: Token(Identifier, "y", (20,2)->(20,3)),
                                },
                            ),
                            colon: Token(SpacedColon, ": ", (20,3)->(20,5)),
                            value: Identifier(
                                Raw(y),
                            ),
                            comma: Some(
                                Token(Comma, ",", (20,6)->(20,7)),
                            ),
                        },
                    ],
                    close: Token(RightBrace, "}", (25,0)->(25,1)),
                },
            ),
            eol: Token(EndOfLine, "\n", (25,1)->(25,1)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (26,0)->(26,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (27,0)->(27,3)),
            ty: None,
            name: Raw(keyed),
            equal: Token(Equal, "=", (27,10)->(27,11)),
            expr: Dict(
                DictLiteral {
                    open: Token(LeftBrace, "{", (27,12)->(27,13)),
                    elements: [
                        KeyValue {
                            key: Expression(
                                String(
                                    DoubleQuote(
                                        "hello",
                                    ),
                                ),
                            ),
                            colon: Token(SpacedColon, ": ", (28,13)->(28,15)),
                            value: String(
                                DoubleQuote(
                                    "world",
                                ),
                            ),
                            comma: Some(
                                Token(Comma, ",", (28,22)->(28,23)),
                            ),
                        },
                        KeyValue {
                            key: Expression(
                                String(
                                    Interpolated(
                                        "{bnr}",
                                    ),
                                ),
                            ),
                            colon: Token(SpacedColon, ": ", (29,14)->(29,16)),
                            value: Identifier(
                                Raw(diag_by_lnum),
                            ),
                            comma: Some(
                                Token(Comma, ",", (29,28)->(29,29)),
                            ),
                        },
                    ],
                    close: Token(RightBrace, "}", (30,0)->(30,1)),
                },
            ),
            eol: Token(EndOfLine, "\n", (30,1)->(30,1)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (31,0)->(31,0)),
    ),
    NoOp(
        Token(EndOfLine, "\n", (32,0)->(32,0)),
    ),
    Call(
        CallCommand {
            call: None,
            expr: Identifier(
                Raw(prop_add),
            ),
            open: Token(LeftParen, "(", (33,8)->(33,9)),
            args: [
                Call(
                    f: Identifier(Raw(line)) arg: [
                        String(
                            SingleQuote(
                                ".",
                            ),
                        ),
                    ],
                ),
                Identifier(
                    Raw(col),
                ),
                Dict(
                    DictLiteral {
                        open: Token(LeftBrace, "{", (33,25)->(33,26)),
                        elements: [
                            KeyValue {
                                key: Literal(
                                    Literal {
                                        token: Token(Identifier, "type", (33,26)->(33,30)),
                                    },
                                ),
                                colon: Token(SpacedColon, ": ", (33,30)->(33,32)),
                                value: String(
                                    SingleQuote(
                                        "LspOutlineHighlight",
                                    ),
                                ),
                                comma: Some(
                                    Token(Comma, ",", (33,53)->(33,54)),
                                ),
                            },
                            KeyValue {
                                key: Literal(
                                    Literal {
                                        token: Token(Identifier, "length", (34,10)->(34,16)),
                                    },
                                ),
                                colon: Token(SpacedColon, ": ", (34,16)->(34,18)),
                                value: MethodCall(
                                    MethodCall {
                                        left: DictAccess(
                                            DictAccess {
                                                container: Index(
                                                    IndexExpression {
                                                        container: DictAccess(
                                                            DictAccess {
                                                                container: Identifier(
                                                                    Scope(ScopedIdentifier { scope: Window, colon: Token(Colon, ":", (34,19)->(34,20)), accessor: Raw(lspSymbols) }),
                                                                ),
                                                                dot: Token(Dot, ".", (34,30)->(34,31)),
                                                                index: RawIdentifier {
                                                                    name: "lnumTable",
                                                                },
                                                            },
                                                        ),
                                                        open: Token(LeftBracket, "[", (34,40)->(34,41)),
                                                        index: Item(
                                                            Identifier(
                                                                Raw(lnum),
                                                            ),
                                                        ),
                                                        close: Token(RightBracket, "]", (34,45)->(34,46)),
                                                    },
                                                ),
                                                dot: Token(Dot, ".", (34,46)->(34,47)),
                                                index: RawIdentifier {
                                                    name: "name",
                                                },
                                            },
                                        ),
                                        tok: Token(MethodArrow, "->", (34,51)->(34,53)),
                                        right: f: Identifier(Raw(len)) arg: [],
                                    },
                                ),
                                comma: None,
                            },
                        ],
                        close: Token(RightBrace, "}", (34,58)->(34,59)),
                    },
                ),
            ],
            close: Token(RightParen, ")", (34,59)->(34,60)),
            eol: Token(EndOfLine, "\n", (34,60)->(34,60)),
        },
    ),
]
